File: 61-rotate-list/61-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(k == 0 || head == null || head.next == null) return head;
        int size = 0;
        for(ListNode curr = head; curr != null; curr = curr.next) size++;
        k %= size;
        ListNode f = head, s = head;
        for(int i = 0; i < k; i++) f = f.next;
        while(f.next != null) {
            f = f.next;
            s = s.next;
        }
        f.next = head; 
        head = s.next;
        s.next = null;
        return head;
        
    }
}

File: 61-rotate-list/61-rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or not head.next: return head
        n = 0
        curr = head 
        while curr:
            curr = curr.next
            n += 1
        k = k % n
        first, second = head, head
        for i in range(k):
            first = first.next
        while first.next:
            first = first.next
            second = second.next
        
        first.next = head
        newHead = second.next
        second.next = None
        return newHead